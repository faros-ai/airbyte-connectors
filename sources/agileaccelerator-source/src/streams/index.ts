import {Works} from './works';

export {Works};
