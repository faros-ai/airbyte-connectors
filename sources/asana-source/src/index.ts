import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {AirbyteSourceLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Asana, AsanaConfig} from './asana';
import {
  Projects,
  ProjectTasks,
  Tags,
  Tasks,
  Users,
  Workspaces,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new AsanaSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Asana source implementation. */
export class AsanaSource extends AirbyteSourceBase<AsanaConfig> {
  get type(): string {
    return 'asana';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AsanaConfig): Promise<[boolean, VError]> {
    try {
      const asana = Asana.instance(config, this.logger);
      await asana.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AsanaConfig): AirbyteStreamBase[] {
    return [
      new Projects(config, this.logger),
      new Tags(config, this.logger),
      new Tasks(config, this.logger),
      new Users(config, this.logger),
      new Workspaces(config, this.logger),
      new ProjectTasks(config, this.logger),
    ];
  }
}
