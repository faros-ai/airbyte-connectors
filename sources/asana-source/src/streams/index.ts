import {ProjectTasks} from './project_tasks';
import {Projects} from './projects';
import {Tags} from './tags';
import {Tasks} from './tasks';
import {Users} from './users';
import {Workspaces} from './workspaces';

export {Projects, ProjectTasks, Tags, Tasks, Users, Workspaces};
