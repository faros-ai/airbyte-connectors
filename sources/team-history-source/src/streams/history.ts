import {AirbyteLogger, AirbyteStreamBase, StreamKey} from 'faros-airbyte-cdk';
import {FarosClient} from 'faros-js-client';
import fs from 'fs';
import path from 'path';
import {Dictionary} from 'ts-essentials';

function loadQueryFile(query: string): string {
  return fs.readFileSync(
    path.join(__dirname, '..', '..', 'resources', 'queries', query),
    'utf8'
  );
}

const CURRENT_TEAM_MEMBERSHIPS_QUERY = loadQueryFile(
  'current-team-memberships.gql'
);
const PREVIOUS_TEAM_MEMBERSHIPS_QUERY = loadQueryFile(
  'previous-team-memberships.gql'
);

interface TeamMembership {
  readonly teamUid: string;
  readonly memberUid: string;
  readonly startedAt?: Date;
}

export class TeamMembershipHistory extends AirbyteStreamBase {
  constructor(
    private readonly faros: FarosClient,
    private readonly graph: string,
    protected readonly logger: AirbyteLogger
  ) {
    super(logger);
  }

  getJsonSchema(): Dictionary<any, string> {
    return require('../../resources/schemas/team-membership-history.json');
  }

  get primaryKey(): StreamKey {
    return ['teamUid', 'memberUid', 'startedAt'];
  }

  private async *readCurrentTeamMemberships(): AsyncGenerator<TeamMembership> {
    const records = this.faros.nodeIterable(
      this.graph,
      CURRENT_TEAM_MEMBERSHIPS_QUERY
    );

    for await (const {team, member} of records) {
      if (team?.uid && member?.uid) {
        yield {teamUid: team.uid, memberUid: member.uid};
      }
    }
  }

  private async *readPreviousTeamMemberships(): AsyncGenerator<TeamMembership> {
    const records = this.faros.nodeIterable(
      this.graph,
      PREVIOUS_TEAM_MEMBERSHIPS_QUERY
    );

    for await (const {team, member, startedAt} of records) {
      if (team?.uid && member?.uid && startedAt) {
        yield {
          teamUid: team.uid,
          memberUid: member.uid,
          startedAt: new Date(startedAt),
        };
      }
    }
  }

  async *readRecords(): AsyncGenerator<Dictionary<any, string>> {
    const prevTeamsByEmployee = new Map<string, Map<string, Date>>();
    const prevTeamMemberships = this.readPreviousTeamMemberships();
    for await (const {teamUid, memberUid, startedAt} of prevTeamMemberships) {
      if (!prevTeamsByEmployee.has(memberUid)) {
        prevTeamsByEmployee.set(memberUid, new Map<string, Date>());
      }
      prevTeamsByEmployee.get(memberUid)?.set(teamUid, startedAt);
    }

    const teamsByEmployee = new Map<string, Set<string>>();
    const currentTeamMemberships = this.readCurrentTeamMemberships();
    for await (const {teamUid, memberUid} of currentTeamMemberships) {
      if (!teamsByEmployee.has(memberUid)) {
        teamsByEmployee.set(memberUid, new Set());
      }
      teamsByEmployee.get(memberUid)?.add(teamUid);
    }

    // Use a fixed timestamp so that employees who switched teams have a
    // contiguous transition from the old team to the new team
    const now = new Date();

    // This loop checks for team memberships that have ended since last time
    for (const [memberUid, prevTeams] of prevTeamsByEmployee) {
      for (const [teamUid, startedAt] of prevTeams) {
        if (!teamsByEmployee.get(memberUid)?.has(teamUid)) {
          yield {
            teamUid,
            memberUid,
            startedAt,
            endedAt: now,
          };
        }
      }
    }

    // This loop checks for team memberships that have started since last time
    for (const [memberUid, teamUids] of teamsByEmployee) {
      for (const teamUid of teamUids) {
        if (!prevTeamsByEmployee.get(memberUid)?.has(teamUid)) {
          yield {teamUid, memberUid, startedAt: now};
        }
      }
    }
  }
}
