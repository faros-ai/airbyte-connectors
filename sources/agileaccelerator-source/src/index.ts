import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  Agileaccelerator,
  AgileacceleratorConfig,
} from './agileaccelerator/agileaccelerator';
import {Works} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AgileacceleratorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AgileAccelerator source implementation. */
export class AgileacceleratorSource extends AirbyteSourceBase<AgileacceleratorConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AgileacceleratorConfig
  ): Promise<[boolean, VError]> {
    try {
      const agileaccelerator = await Agileaccelerator.instance(
        config,
        this.logger
      );
      await agileaccelerator.checkConnection(config);
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AgileacceleratorConfig): AirbyteStreamBase[] {
    return [new Works(config, this.logger)];
  }
}
