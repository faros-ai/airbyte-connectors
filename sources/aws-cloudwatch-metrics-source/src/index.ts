import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {CloudWatch, Config} from './cloudwatch';
import {Metrics} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new CloudWatchMetricsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** CloudWatch Metrics source implementation. */
export class CloudWatchMetricsSource extends AirbyteSourceBase<Config> {
  get type(): string {
    return 'cloudwatch-metrics';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: Config): Promise<[boolean, VError]> {
    try {
      const cloudWatch = CloudWatch.instance(config);
      await cloudWatch.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: Config): AirbyteStreamBase[] {
    return [new Metrics(config, this.logger)];
  }
}
