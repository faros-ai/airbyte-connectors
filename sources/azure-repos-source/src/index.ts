import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureRepoConfig, AzureRepos} from './azure-repos';
import {PullRequests, Repositories, Users} from './streams';
import {Commits} from './streams/commits';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureRepoSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzureRepo source implementation. */
export class AzureRepoSource extends AirbyteSourceBase<AzureRepoConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AzureRepoConfig): Promise<[boolean, VError]> {
    try {
      const azureRepos = await AzureRepos.make(config, this.logger);
      await azureRepos.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureRepoConfig): AirbyteStreamBase[] {
    return [
      new Commits(config, this.logger),
      new PullRequests(config, this.logger),
      new Repositories(config, this.logger),
      new Users(config, this.logger),
    ];
  }
}
