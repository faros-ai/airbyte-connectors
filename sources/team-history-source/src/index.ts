import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {TeamMembershipHistory} from './streams';
import {FarosClient} from 'faros-js-client';

export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new TeamHistorySource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

interface TeamHistorySourceConfig {
  readonly api_url: string;
  readonly api_key: string;
  readonly graph: string;
}

export class TeamHistorySource extends AirbyteSourceBase<TeamHistorySourceConfig> {
  get type(): string {
    return 'team-history';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }

  async checkConnection(
    cfg: TeamHistorySourceConfig
  ): Promise<[boolean, VError]> {
    try {
      const faros = this.makeFarosClient(cfg);
      if (!(await faros.graphExists(cfg.graph))) {
        return [false, new VError(`Graph ${cfg.graph} does not exist!`)];
      }
    } catch (err: any) {
      return [false, err as VError];
    }
    return [true, undefined];
  }

  streams(cfg: TeamHistorySourceConfig): AirbyteStreamBase[] {
    const faros = this.makeFarosClient(cfg);
    return [new TeamMembershipHistory(faros, cfg.graph, this.logger)];
  }

  private makeFarosClient(cfg: TeamHistorySourceConfig): FarosClient {
    if (!cfg.api_key) throw new VError('Faros API key was not provided');
    if (!cfg.graph) throw new VError('Faros graph name was not provided');
    return new FarosClient({
      url: cfg.api_url,
      apiKey: cfg.api_key,
    });
  }
}
