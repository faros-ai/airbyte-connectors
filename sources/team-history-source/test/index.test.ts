import {
  AirbyteLogLevel,
  AirbyteSourceLogger,
  AirbyteSpec,
} from 'faros-airbyte-cdk';
import {
  readResourceAsJSON,
  readTestResourceAsJSON,
  sourceCheckTest,
  sourceReadTest,
} from 'faros-airbyte-testing-tools';
import {FarosClient} from 'faros-js-client';
import VError from 'verror';

import * as sut from '../src/index';
import {TeamMembershipHistory} from '../src/streams';

interface TeamMembershipModel {
  readonly team: {uid: string} | null;
  readonly member: {uid: string} | null;
  readonly startedAt?: string;
}

interface MockGraphQueries {
  readonly currentMemberships?: TeamMembershipModel[];
  readonly previousMemberships?: TeamMembershipModel[];
}

function mockGraphQueries(mock: MockGraphQueries): void {
  jest
    .spyOn(FarosClient.prototype, 'nodeIterable')
    .mockImplementation((_graph, query) => {
      if (query.includes('CurrentTeamMemberships')) {
        return mockAsyncIterable(mock.currentMemberships ?? []);
      } else if (query.includes('PreviousTeamMemberships')) {
        return mockAsyncIterable(mock.previousMemberships ?? []);
      }
      return mockAsyncIterable([]);
    });
}

describe('index', () => {
  const logger = new AirbyteSourceLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.FATAL
  );

  const source = new sut.TeamHistorySource(logger);

  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('spec', async () => {
    await expect(source.spec()).resolves.toStrictEqual(
      new AirbyteSpec(readResourceAsJSON('spec.json'))
    );
  });

  describe('check connection', () => {
    test('valid', async () => {
      jest.spyOn(FarosClient.prototype, 'graphExists').mockResolvedValue(true);

      await sourceCheckTest({
        source,
        configOrPath: 'check_connection/valid_config.json',
      });
    });

    test('missing graph', async () => {
      await sourceCheckTest({
        source,
        configOrPath: 'check_connection/missing_graph.json',
      });
    });

    test('graph does not exist', async () => {
      jest.spyOn(FarosClient.prototype, 'graphExists').mockResolvedValue(false);

      await sourceCheckTest({
        source,
        configOrPath: 'check_connection/valid_config.json',
      });
    });

    test('api error', async () => {
      jest
        .spyOn(FarosClient.prototype, 'graphExists')
        .mockRejectedValue(new VError('API connection failed'));

      await sourceCheckTest({
        source,
        configOrPath: 'check_connection/valid_config.json',
      });
    });
  });

  test('streams', () => {
    const config = readTestResourceAsJSON('valid_config.json');
    const streams = source.streams(config);

    expect(streams).toHaveLength(1);
    expect(streams[0]).toBeInstanceOf(TeamMembershipHistory);
  });

  describe('TeamMembershipHistory stream', () => {
    test('schema', () => {
      const config = readTestResourceAsJSON('valid_config.json');
      const farosClient = new FarosClient({
        url: config.api_url,
        apiKey: config.api_key,
      });
      const stream = new TeamMembershipHistory(
        farosClient,
        config.graph,
        logger
      );
      const schema = stream.getJsonSchema();
      expect(schema).toMatchSnapshot();
    });

    test('primary key', () => {
      const config = readTestResourceAsJSON('valid_config.json');
      const farosClient = new FarosClient({
        url: config.api_url,
        apiKey: config.api_key,
      });
      const stream = new TeamMembershipHistory(
        farosClient,
        config.graph,
        logger
      );
      expect(stream.primaryKey).toEqual(['teamUid', 'memberUid', 'startedAt']);
    });

    test('readRecords - new memberships', async () => {
      mockGraphQueries({
        currentMemberships: [
          {team: {uid: 'team1'}, member: {uid: 'user1'}},
          {team: {uid: 'team2'}, member: {uid: 'user1'}},
          {team: {uid: 'team1'}, member: {uid: 'user2'}},
        ],
        previousMemberships: [
          {
            team: {uid: 'team1'},
            member: {uid: 'user1'},
            startedAt: '2023-01-01T00:00:00Z',
          },
        ],
      });

      await sourceReadTest({
        source,
        configOrPath: 'valid_config.json',
        catalogOrPath: 'catalog.json',
        checkRecordsData: (records) => {
          expect(records).toHaveLength(2);
          expect(records).toContainEqual(
            expect.objectContaining({
              teamUid: 'team2',
              memberUid: 'user1',
              startedAt: expect.any(Date),
            })
          );

          expect(records).toContainEqual(
            expect.objectContaining({
              teamUid: 'team1',
              memberUid: 'user2',
              startedAt: expect.any(Date),
            })
          );
        },
      });
    });

    test('readRecords - updated memberships', async () => {
      mockGraphQueries({
        currentMemberships: [{team: {uid: 'team1'}, member: {uid: 'user1'}}],
        previousMemberships: [
          {
            team: {uid: 'team1'},
            member: {uid: 'user1'},
            startedAt: '2023-01-01T00:00:00Z',
          },
          {
            team: {uid: 'team2'},
            member: {uid: 'user1'},
            startedAt: '2023-02-01T00:00:00Z',
          },
        ],
      });

      await sourceReadTest({
        source,
        configOrPath: 'valid_config.json',
        catalogOrPath: 'catalog.json',
        checkRecordsData: (records) => {
          expect(records).toHaveLength(1);
          expect(records).toContainEqual(
            expect.objectContaining({
              memberUid: 'user1',
              teamUid: 'team2',
              startedAt: new Date('2023-02-01T00:00:00Z'),
              endedAt: expect.any(Date),
            })
          );
        },
      });
    });

    test('readRecords - no changes', async () => {
      mockGraphQueries({
        currentMemberships: [{team: {uid: 'team1'}, member: {uid: 'user1'}}],
        previousMemberships: [
          {
            team: {uid: 'team1'},
            member: {uid: 'user1'},
            startedAt: '2023-01-01T00:00:00Z',
          },
        ],
      });

      await sourceReadTest({
        source,
        configOrPath: 'valid_config.json',
        catalogOrPath: 'catalog.json',
        checkRecordsData: (records) => {
          expect(records).toHaveLength(0);
        },
      });
    });

    test('readRecords - ignore records with nulls', async () => {
      mockGraphQueries({
        currentMemberships: [
          {team: {uid: 'team1'}, member: {uid: 'user1'}},
          {team: null, member: {uid: 'user2'}}, // Missing team
          {team: {uid: 'team2'}, member: null}, // Missing member
          {team: {uid: null}, member: {uid: 'user3'}}, // Missing team uid
          {team: {uid: 'team3'}, member: {uid: null}}, // Missing member uid
        ],
        previousMemberships: [
          {
            team: {uid: 'team1'},
            member: {uid: 'user1'},
            startedAt: '2023-01-01T00:00:00Z',
          },
          {team: {uid: 'team4'}, member: {uid: 'user4'}, startedAt: null}, // Missing startedAt
        ],
      });

      await sourceReadTest({
        source,
        configOrPath: 'valid_config.json',
        catalogOrPath: 'catalog.json',
        checkRecordsData: (records) => {
          expect(records).toHaveLength(0);
        },
      });
    });
  });
});

// Helper function to create async iterable from array
async function* mockAsyncIterable<T>(items: T[]): AsyncGenerator<T> {
  for (const item of items) {
    yield item;
  }
}
