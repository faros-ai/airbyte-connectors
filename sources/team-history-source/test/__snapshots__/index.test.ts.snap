// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index TeamMembershipHistory stream schema 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "properties": {
    "endedAt": {
      "format": "date-time",
      "type": [
        "string",
        "null",
      ],
    },
    "memberUid": {
      "type": "string",
    },
    "startedAt": {
      "format": "date-time",
      "type": "string",
    },
    "teamUid": {
      "type": "string",
    },
  },
  "required": [
    "teamUid",
    "memberUid",
    "startedAt",
  ],
  "type": "object",
}
`;

exports[`index check connection api error 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "API connection failed",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection graph does not exist 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Graph test-graph does not exist!",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection missing graph 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "Faros graph name was not provided",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`index check connection valid 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "status": "SUCCEEDED",
  },
  "type": "CONNECTION_STATUS",
}
`;
